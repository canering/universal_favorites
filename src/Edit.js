import React from "react";
import PropTypes from "prop-types";

class Edit extends React.Component {

    constructor(props){
        super(props);

        let isFolder = false;
        let url = "";
        if(props.link.type === "folder") {
            isFolder = true;
        }
        else {
            url = props.link.link;
        }

        this.state = {
            name:props.link.label,
            url:url,
            isFolder:isFolder
        };
    }

    static get propTypes(){
        return({
            goBack:PropTypes.func,
            save:PropTypes.func,
            link:PropTypes.object
        });
    }

    render(){
        return(
            <div className="menu">
                <button id="onlyBack" className="menuButton" onClick={this.backToFavs}>
                    <img src="https://img.icons8.com/ios-filled/50/000000/back.png" alt="back" className="icon" />
                </button><br />

				Name: <input type="text" onChange={this.updateName} value={this.state.name} /><br />
                {this.state.isFolder ? 
                    null
                    : <React.Fragment>URL: <input type="text" onChange={this.updateUrl} value={this.state.url} /><br /></React.Fragment>}

                <button onClick={this.submit}>Submit</button>
            </div>
        );
    }

updateName = (event) => {
    this.setState({name:event.target.value});
}
updateUrl = (event) => {
    this.setState({url:event.target.value});
}
backToFavs = () => {
    this.props.goBack();
}
submit = () => {
    if(this.state.name === "") {
        return;
    }

    let link = Object.assign({}, this.props.link);

    link.label = this.state.name;

    if(this.state.isFolder){
        this.props.save(link);
    }
    else{
        if(this.state.url === "") {
            return;
        }

        link.link = this.state.url;

        this.props.save(link);
    }
}
}

export default Edit;