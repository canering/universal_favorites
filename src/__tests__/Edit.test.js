import React from "react";
import {shallow, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import Edit from "../Edit";

configure({adapter: new Adapter()});

const linkProps = {
    goBack: jest.fn(),
    save: jest.fn(),
    link: {type:"link", link:"http://example.com", label:"example.com"}
};

const folderProps = {
    goBack: jest.fn(),
    save: jest.fn(),
    link: {type:"folder", label:"example.com"}
};

afterEach(() => {
    jest.clearAllMocks();
});

test("edit link shapshot test", () => {
    const component = renderer.create(<Edit {...linkProps} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("edit folder shapshot test", () => {
    const component = renderer.create(<Edit {...folderProps} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("when back is clicked than goBack should be called", () => {
    const component = shallow(<Edit {...linkProps} />);
    component.find("#onlyBack").prop("onClick")();

    expect(linkProps.goBack).toHaveBeenCalledTimes(1);
});

test("when folder is changed then save should be called", () => {
    const component = shallow(<Edit {...folderProps} />);
    component.find("input").at(0).simulate("change", {target: {value: "example2.com"}});
    component.find(".plainButton").prop("onClick")();

    expect(folderProps.save).toHaveBeenCalledTimes(1);
    expect(folderProps.save).toHaveBeenCalledWith({type:"folder", label:"example2.com"});
});

test("when link is changed then save should be called", () => {
    const component = shallow(<Edit {...linkProps} />);
    const inputs = component.find("input");
    inputs.at(0).simulate("change", {target: {value: "example2.com"}});
    inputs.at(1).simulate("change", {target: {value: "http://example2.com"}});

    component.find(".plainButton").prop("onClick")();

    expect(linkProps.save).toHaveBeenCalledTimes(1);
    expect(linkProps.save).toHaveBeenCalledWith({type:"link", link:"http://example2.com", label:"example2.com"});
});