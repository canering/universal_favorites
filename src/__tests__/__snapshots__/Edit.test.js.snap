// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit folder shapshot test 1`] = `
<div
  className="menu"
>
  <button
    className="menuButton"
    id="onlyBack"
    onClick={[Function]}
  >
    <img
      alt="back"
      className="icon"
      src="back.png"
    />
  </button>
  <br />
  Name: 
  <input
    onChange={[Function]}
    type="text"
    value="example.com"
  />
  <br />
  <button
    className="plainButton"
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;

exports[`edit link shapshot test 1`] = `
<div
  className="menu"
>
  <button
    className="menuButton"
    id="onlyBack"
    onClick={[Function]}
  >
    <img
      alt="back"
      className="icon"
      src="back.png"
    />
  </button>
  <br />
  Name: 
  <input
    onChange={[Function]}
    type="text"
    value="example.com"
  />
  <br />
  URL: 
  <input
    onChange={[Function]}
    type="text"
    value="http://example.com"
  />
  <br />
  <button
    className="plainButton"
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;
